soccer/KickEvaluator.cpp

```diff
@@ -0,0 +1,396 @@
+    static float derivative;
+
+    sterm =
+        sqrtpi_2 * (sqrt1_kstdev2 - kstdev * erf((kx + bLeft) / sqrt2_kstdev));
+
+    tterm =
+        sqrtpi_2 * (sqrt1_kstdev2 - kstdev * erf((kx + bRight) / sqrt2_kstdev));
+
+    // For each robot distribution in Set #1
+    static int i;
+    for (i = 0; i < robotMeans.size(); i++) {
+        rmean = robotMeans[i];
+        rstdev = robotStDevs[i];
+        rstdev2 = rstdev * rstdev;
+        robotV = robotVertScales[i];
+
+        fterm = -1.0f * fast_exp(-0.5f * (kx + rmean) * (kx + rmean) /
+                                 (kstdev2 + rstdev2)) *
+                robotV * sqrt2pi;
+        fterm = fterm / sqrt(1.0f / kstdev2 + 1.0f / rstdev2);
+
+        results = 1.0f / (kstdev * sqrt2pi) * (fterm + sterm - tterm);
+
+        if (results < minResults) {
+            minResults = results;
+            minIndex = i;
+        }
+    }
+
+    // Calculate derivative of the convolution
+    rmean = robotMeans[minIndex];
+    rstdev = robotStDevs[minIndex];
+    rstdev2 = rstdev * rstdev;
+    robotV = robotVertScales[minIndex];
+
+    fterm =
+        fast_exp(-0.5f * (kx + rmean) * (kx + rmean) / (kstdev2 + rstdev2)) *
+        robotV * sqrt2pi * (kx + rmean);
+    fterm =
+        fterm / (sqrt(1.0f / kstdev2 + 1.0f / rstdev2) * (kstdev2 + rstdev2));
+
+    sterm = fast_exp(-0.5f * (kx + bLeft) * (kx + bLeft) / kstdev2);
+
+    tterm = fast_exp(-0.5f * (kx + bRight) * (kx + bRight) / kstdev2);
+
+    derivative = 1.0f / (kstdev * sqrt2pi) * (sterm - tterm - fterm);
+
+    return forward_as_tuple(minResults, derivative);
+}
+
+float KickEvaluator::get_target_angle(const Point origin,
+                                      const Segment target) {
+    Point left = target.pt[0] - origin;
+    Point right = target.pt[1] - origin;
+
+    return abs(left.angle() - right.angle());
+}
+
+vector<Robot*> KickEvaluator::get_valid_robots() {
+    vector<Robot*> bots(system->self.size() + system->opp.size());
+
+    auto filter_predicate = [&](const Robot* bot) -> bool {
+        return bot != nullptr && bot->visible &&
+               find(excluded_robots.begin(), excluded_robots.end(), bot) ==
+                   excluded_robots.end();
+    };
+
+    auto end_it = copy_if(system->self.begin(), system->self.end(),
+                          bots.begin(), filter_predicate);
+
+    end_it = copy_if(system->opp.begin(), system->opp.end(), end_it,
+                     filter_predicate);
+
+    bots.resize(distance(bots.begin(), end_it));
+
+    return bots;
+}
+
+tuple<float, float> KickEvaluator::rect_to_polar(const Point origin,
+                                                 const Point target,
+                                                 const Point obstacle) {
+    Point obstacleDir = obstacle - origin;
+    Point targetDir = target - origin;
+    float angle = obstacleDir.angle() - targetDir.angle();
+
+    return forward_as_tuple(obstacleDir.mag(), fixAngleRadians(angle));
```

Comment:
It returns a tuple<float &&, float&&>. I played around with it a bit and they're about the same speed. I'm gonna switch it to make_tuple since that makes more sense in context